confgenerator/feature.go

```diff
@@ -0,0 +1,176 @@
+package confgenerator
+
+import (
+	"errors"
+	"fmt"
+	"reflect"
+	"strings"
+)
+
+var ErrTrackingInlineStruct = errors.New("cannot have tracking on inline struct")
+
+type Feature struct {
+	Module, Kind, Type, Key, Value string
+}
+
+func ExtractFeatures(uc *UnifiedConfig) ([]Feature, error) {
+	allFeatures := getOverriddenDefaultPipelines(uc)
+
+	if uc.Metrics != nil {
+		for _, r := range uc.Metrics.Receivers {
+			features, err := trackingFeatures(r, MetricsReceiverTypes.Subagent, MetricsReceiverTypes.Kind, r.Type())
+			if err != nil {
+				return nil, err
+			}
+			allFeatures = append(allFeatures, features...)
+		}
+		for _, p := range uc.Metrics.Processors {
+			features, err := trackingFeatures(p, MetricsProcessorTypes.Subagent, MetricsProcessorTypes.Kind, p.Type())
+			if err != nil {
+				return nil, err
+			}
+			allFeatures = append(allFeatures, features...)
+		}
+	}
+	if uc.Logging != nil {
+		for _, r := range uc.Logging.Receivers {
+			features, err := trackingFeatures(r, LoggingReceiverTypes.Subagent, LoggingReceiverTypes.Kind, r.Type())
+			if err != nil {
+				return nil, err
+			}
+			allFeatures = append(allFeatures, features...)
+		}
+		for _, p := range uc.Logging.Processors {
+			features, err := trackingFeatures(p, LoggingProcessorTypes.Subagent, LoggingProcessorTypes.Kind, p.Type())
+			if err != nil {
+				return nil, err
+			}
+			allFeatures = append(allFeatures, features...)
+		}
+	}
+
+	return allFeatures, nil
+}
+
+func trackingFeatures(component interface{}, module, kind, typ string) ([]Feature, error) {
+	return trackingFeaturesHelper(component, module, kind, typ, "")
+}
+
+func trackingFeaturesHelper(component any, module, kind, typ, prefix string) ([]Feature, error) {
+	if component == nil {
+		return []Feature{}, nil
+	}
+	c := reflect.ValueOf(component)
+	t := c.Type()
+
+	if t.Kind() == reflect.Ptr {
+		t = t.Elem()
+	}
+
+	var features []Feature
+	// Iterate over all available fields and read the tag value
+	for i := 0; i < t.NumField(); i++ {
+		// Get the field, returns https://golang.org/pkg/reflect/#StructField
+		field := t.Field(i)
+
+		// Grab actual value instead of pointer
+		v := reflect.Indirect(reflect.Indirect(c).FieldByName(field.Name))
+		if v.IsZero() {
```

Comment:
I am not sure about `v.IsZero()` here - my understanding is `false` is the zero value for bool? If the value we want to track is bool and is false, we would skip it incorrectly, right?

Reply:
The zero value of a feature can be inferred and therefore it does not need to be tracked.